rust/lance/src/dataset.rs

```diff
@@ -439,21 +441,53 @@ impl Dataset {
         batches: impl RecordBatchReader + Send + 'static,
         dest: impl Into<WriteDestination<'_>>,
         params: Option<WriteParams>,
+    ) -> Result<Self> {
+        let (batches, schema) = peek_reader_schema(Box::new(batches)).await?;
+        let stream = reader_to_stream(batches);
```

Comment:
We do some important logic related to dictionary arrays in `peek_reader_schema`. So I think we should take `Schema` from the user in the stream case. Instead, we should flip this around:

```rust
        let stream = reader_to_stream(batches);
        let (stream, schema) = peek_stream_schema(stream).await?;
```

Then for the stream case we can also use the `peek_stream_schema()` method.

Reply:
Implemented by `StreamingWriteSource::into_stream_and_schema`. People who do not want to get schema can rather use `StreamingWriteSource::into_stream`, which is the formal `reader_to_stream`. Of cource, `reader_to_stream` function would not removed.